Mathlib/Tactic/SolveByElim.lean

```diff
@@ -15,40 +18,291 @@ We'll gradually bring it up to feature parity.
 
 open Lean Meta Elab Tactic
 
-/-- Return local hypotheses which are not "implementation detail", as `Expr`s. -/
-def Lean.Meta.getLocalHyps : MetaM (Array Expr) := do
-  let mut hs := #[]
-  for d in ← getLCtx do
-    if !d.isImplementationDetail then hs := hs.push d.toExpr
-  return hs
+/-- Visualize an `Except` using a checkmark or a cross. -/
+def exceptEmoji : Except ε α → String
+  | .error _ => crossEmoji
+  | .ok _ => checkEmoji
+
+namespace Lean.MVarId
+
+/--
+`applyFirst lemmas cont goal` will try to apply one of the `lemmas` to the goal `goal`,
+and then call `cont` on the resulting `List MVarId` of subgoals.
+
+It returns the result from `cont` for the first such lemma for which
+both the `apply` and the call to `cont` succeed.
+
+``applyFirst (trace := `name)`` will construct trace nodes for ``name` indicating which
+calls to `apply` succeeded or failed.
+-/
+-- Because the operation of this function via a continuation is fairly specific to `solve_by_elim`,
+-- we keep it here rather than moving it into `Mathlib/Lean/`.
+def applyFirst (cfg : ApplyConfig := {}) (trace : Name := .anonymous) (lemmas : List Expr)
+    (cont : List MVarId → MetaM α) (g : MVarId) : MetaM α :=
+  lemmas.firstM fun e =>
+    withTraceNode trace (return m!"{exceptEmoji ·} tried to apply: {e}") do
+      cont (← g.apply e cfg)
+
+end Lean.MVarId
 
 initialize registerTraceClass `Meta.Tactic.solveByElim
 
 namespace Mathlib.Tactic.SolveByElim
 
+/--
+Configuration structure to control the behaviour of `solve_by_elim`:
+*  control the maximum depth and behaviour (fail or return subgoals) at the maximum depth
+*  whether to use `symm` on hypotheses and `exfalso` on the goal as needed,
+* and hooks allowing
+  * modifying intermediate goals,
+  * returning goals as subgoals, and
+  * discharging subgoals.
+-/
+structure Config extends ApplyConfig where
+  /-- Maximum recursion depth. -/
+  maxDepth : Nat := 12
```

Comment:
The [corresponding default max-depth](https://github.com/leanprover-community/mathlib/blob/c982179ec21091d3e102d8a5d9f5fe06c8fafb73/src/tactic/solve_by_elim.lean#L126) in mathlib3 is 3, if I understand correctly.

What's the reason for making it so much larger here? Could this potentially make some proofs go significantly slower than this did in mathlib3?

Reply:
Yes, that was excessive. I've been wanting to increase it in mathlib for a long time. I mistakenly thought it was 6 in mathlib3, and thought I was doubling it. :-)

Raising it to 6 in mathlib3 seems to not cause any problems (any side goal from a `wlog` is not longer needed), so I propose we try that here as a compromise!